### PR TITLE
[SPARK-28004][UI] Update jquery to 3.4.1

### PR DESCRIPTION
**Pull request by srowen**
## What changes were proposed in this pull request?

We're using an old-ish jQuery, 1.12.4, and should probably update for Spark 3 to keep up in general, but also to keep up with CVEs. In fact, we know of at least one resolved in only 3.4.0+ (https://nvd.nist.gov/vuln/detail/CVE-2019-11358). They may not affect Spark, but, if the update isn't painful, maybe worthwhile in order to make future 3.x updates easier.

jQuery 1 -> 2 doesn't sound like a breaking change, as 2.0 is supposed to maintain compatibility with 1.9+ (https://blog.jquery.com/2013/04/18/jquery-2-0-released/)

2 -> 3 has breaking changes: https://jquery.com/upgrade-guide/3.0/. It's hard to evaluate each one, but the most likely area for problems is in ajax(). However, our usage of jQuery (and plugins) is pretty simple.

Update jquery to 3.4.1; update jquery blockUI and mustache to latest

## How was this patch tested?

Manual testing of docs build (except R docs), worker/master UI, spark application UI.
Note: this really doesn't guarantee it works, as our tests can't test javascript, and this is merely anecdotal testing, although I clicked about every link I could find. There's a risk this breaks a minor part of the UI; it does seem to work fine in the main.

### CODE REVIEW

File: core/src/main/resources/org/apache/spark/ui/static/jquery.mustache.js [hunk 1]
```diff
@@ -45,6 +45,19 @@
     return obj != null && typeof obj === 'object' && (propName in obj);
   }
 
+  /**
```
Comment (srowen) [inline core/src/main/resources/org/apache/spark/ui/static/jquery.mustache.js]: These changes are all from the source jquery.mustache.js, not custom modifications

File: docs/js/main.js [hunk 2]
```diff
@@ -93,5 +93,5 @@ $(function() {
 
   // Scroll now too in case we had opened the page on a hash, but wait a bit because some browsers
   // will try to do *their* initial scroll after running the onReady handler.
-  $(window).load(function() { setTimeout(function() { maybeScrollToHash(); }, 25); }); 
+  $(window).on('load', function() { setTimeout(function() { maybeScrollToHash(); }, 25); });
```
Comment (srowen) [inline docs/js/main.js]: This is actually the only change I seemed to need to make to the Spark javascript.

**Comment by srowen** [comment]
Ah right. The history server. There's a problem there. I'll look into it.

**Comment by srowen** [comment]
Hm, for anyone following along, these tests pass locally. Maybe I'm doing something wrong running locally. Looks like an httpclient version mismatch that I'll have to run down: `Cannot locate declared field class : org.apache.http.impl.client.HttpClientBuilder.dnsResolver`

File: core/src/main/resources/org/apache/spark/ui/static/jquery.mustache.js (hunk 1 above)
Comment (dongjoon-hyun) [reply core/src/main/resources/org/apache/spark/ui/static/jquery.mustache.js]: I also checked that. This file is identical with `mustache.js` v3.0.1. Only the trailing spaces are removed.

File: docs/js/main.js (hunk 2 above)
Comment (dongjoon-hyun) [reply docs/js/main.js]: Yep. This is the correct update based on the upgrade-guide.
- https://jquery.com/upgrade-guide/3.0/#breaking-change-load-unload-and-error-removed

File: core/src/main/resources/org/apache/spark/ui/static/historypage.js [hunk 3]
```diff
@@ -104,7 +104,7 @@ $(document).ready(function() {
     });
 
     var historySummary = $("#history-summary");
-    var searchString = historySummary["context"]["location"]["search"];
+    var searchString = window.location.search;
```
Comment (dongjoon-hyun) [inline core/src/main/resources/org/apache/spark/ui/static/historypage.js]: I also verified that this works in the same way.

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM. Thank you, @srowen !

This should be tested in the browser's `private` mode to avoid old javascript cache.
I verified this PR with the followings.
1. `Spark shell application UI`
2. `Thrift JDBC/ODBC Server application UI`
3. `Spark History Server UI` on my several Spark history logs. 

All tabs look working correctly. I might miss a few tiny UI features like `hover` info, but the navigation and the info looks correct in general. I hope we can merge this and proceed.

Merged to master.

**Comment by dongjoon-hyun** [comment]
cc @gatorsmile and @gengliangwang

File: core/src/test/scala/org/apache/spark/deploy/history/HistoryServerSuite.scala [hunk 4]
```diff
@@ -366,9 +367,8 @@ class HistoryServerSuite extends SparkFunSuite with BeforeAndAfter with Matchers
     contextHandler.addServlet(holder, "/")
     server.attachHandler(contextHandler)
 
-    implicit val webDriver: WebDriver = new HtmlUnitDriver(true) {
-      getWebClient.getOptions.setThrowExceptionOnScriptError(false)
-    }
+    implicit val webDriver: WebDriver =
```
Comment (srowen) [inline core/src/test/scala/org/apache/spark/deploy/history/HistoryServerSuite.scala]: One last clarifying comment: This change slightly improves the test to _not_ ignore javascript errors, and as part of that, needed to test as a modern browser (IE 11 vs IE 8).

**Comment by gengliangwang** [comment]
A late LGTM. I also tried `grep -r "jquery-1.12.4.min.js" *` and there is no matched results.

**Comment by gabrieljones** [comment]
any chance this will be back ported to 2.3? 2.3 is currently quarantined at my site because of this.

**Comment by srowen** [comment]
No, 2.3 is EOL. You should update to 2.4.x

**Pull request closed without merging**